File: src/my_cpp_pkg/src/add_two_ints_serv.cpp
#include <memory>

#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

class AddTwoIntsServiceNode : public rclcpp::Node
{
public:
  AddTwoIntsServiceNode()
  : Node("add_two_ints_serv")
  {
    service_ = this->create_service<example_interfaces::srv::AddTwoInts>(
      "add_two_ints",
      std::bind(
        &AddTwoIntsServiceNode::handle_add_two_ints,
        this,
        std::placeholders::_1,
        std::placeholders::_2));

    RCLCPP_INFO(
      this->get_logger(),
      "Service \"add_two_ints\" is ready to receive requests.");
  }

private:
  rclcpp::Service<example_interfaces::srv::AddTwoInts>::SharedPtr service_;

  void handle_add_two_ints(
    const std::shared_ptr<example_interfaces::srv::AddTwoInts::Request> request,
    std::shared_ptr<example_interfaces::srv::AddTwoInts::Response> response)
  {
    response->sum = request->a + request->b;

    RCLCPP_INFO(
      this->get_logger(),
      "Received request: a=%ld, b=%ld. Responding with sum=%ld.",
      request->a,
      request->b,
      response->sum);
  }
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  auto node = std::make_shared<AddTwoIntsServiceNode>();
  rclcpp::spin(node);

  rclcpp::shutdown();
  return 0;
}

File: src/my_py_pkg/my_py_pkg/add_two_ints_serv.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts

class MyNode(Node):
    def __init__(self):
        super().__init__('add_two_ints_serv')
        self.srv = self.create_service(AddTwoInts, 'add_two_ints', self.add_two_ints_callback)
        self.get_logger().info('Service "add_two_ints" is ready to receive requests.')

    def add_two_ints_callback(self, request: AddTwoInts.Request, response: AddTwoInts.Response) -> AddTwoInts.Response:
        response.sum = request.a + request.b
        self.get_logger().info(
            f'Received request: a={request.a}, b={request.b}. Responding with sum={response.sum}.'
        )
        return response
    
def main(args=None):
    rclpy.init(args=args)
    node = MyNode()

    rclpy.spin(node)

    rclpy.shutdown()

if __name__ == '__main__':
    main()
